MVP Seb middleware

Ivanca opened the pull request:
- adding seb_openedx middleware and functions for lms + cms

- seb-info (e.g. http://localhost:18000/seb-openedx/seb-info) now has a middleware_installed key to easily check if the middleware is installed (both cms and lms have such url)

Ivanca committed:
adding seb_openedx middleware and functions for lms + cms

felipemontoya commented on seb_openedx/handle_views.py:
```diff
@@ -0,0 +1,30 @@
+def lms(request, course_module):
+    """ Handle lms view """
+    other_settings = course_module.other_course_settings
+    header = 'HTTP_X_SAFEEXAMBROWSER_CONFIGKEYHASH'
```
Should this be a configuration constant?

Ivanca replied:
I don't think so, that's the only header SEB sets and its not configurable in their end.

felipemontoya commented on seb_openedx/handle_views.py:
```diff
@@ -0,0 +1,30 @@
+    """ Handle lms view """
+    other_settings = course_module.other_course_settings
+    header = 'HTTP_X_SAFEEXAMBROWSER_CONFIGKEYHASH'
+    if 'seb_keys' in other_settings and other_settings['seb_keys']:
```
I dont think that `seb_keys` must be a configuration param, but at least a constant of this module would be nice

felipemontoya commented on seb_openedx/handle_views.py:
```diff
@@ -0,0 +1,30 @@
+    other_settings = course_module.other_course_settings
+    header = 'HTTP_X_SAFEEXAMBROWSER_CONFIGKEYHASH'
+    if 'seb_keys' in other_settings and other_settings['seb_keys']:
+        if request.user.is_staff:
```
Who says staff should be allowed?
This is the kind of thing that I dont even think we should decide. That is why I have talked so much about modular and configurable.

Think of this as a list of conditions (which are configurable) and the function that handles `marking` should go through the list running through each function to determine the final result

felipemontoya commented on seb_openedx/middleware.py:
```diff
@@ -0,0 +1,28 @@
+
+    def process_request(self, request):
+        """ Force this feature to be on (allows usage of "other_course_settings" dict) """
+        settings.FEATURES['ENABLE_OTHER_COURSE_SETTINGS'] = True
```
This is using the django.conf module directly. I think we should get this through a function helper in the plugin so that we can alter the location we use to read the settings in one location. We could have different backends for this

felipemontoya replied:
Also, modifiying the settings.FEATURES dictionary on runtime, specially on a middleware is super uncool. We should not be changing the settings after django has settled them.

felipemontoya commented on seb_openedx/middleware.py:
```diff
@@ -0,0 +1,28 @@
+        settings.FEATURES['ENABLE_OTHER_COURSE_SETTINGS'] = True
+
+    # pylint: disable=inconsistent-return-statements
+    def process_view(self, request, view_func, view_args, view_kwargs):
```
We are using the deprecation.MiddlewareMixin for this to make it compatible with the old format. Am I correct in assuming that process_view is the new way of doing middlewares? or is `__call__` the new way, and this, the compatibility one?

felipemontoya commented on seb_openedx/middleware.py:
```diff
@@ -0,0 +1,28 @@
+            course_key_string = view_kwargs.get('course_key_string') or view_kwargs.get('course_id')
+            course_key = CourseKey.from_string(course_key_string) if course_key_string else None
+            if course_key:
+                course_module = modulestore().get_course(course_key, depth=0)
```
this is a costly operation. It calls the complete mongo stack. We need to find ways to filter this so that only the requests that require this are running this code branch.

felipemontoya commented on seb_openedx/settings/common.py:
```diff
@@ -18,4 +18,4 @@ def plugin_settings(settings):
     See: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/plugins/README.rst
     """
-    pass
+    settings.IS_CMS = settings.__name__.startswith('cms')
```
I think a cleaner solution is to separate the middlewares into a cms middleware and a lms one. We are overcomplicating this a lot, and we are gaining only the ability to auto add a key to the other_configs dict. A user configuring its exam to work on SEB is most definitely a power user and can be trusted to add a specific key to a field in studio.

Ivanca replied:
Done

felipemontoya commented on seb_openedx/utils.py:
```diff
@@ -0,0 +1,11 @@
+import json
+
+
+def parse_json_maybe(string, default):
```
this function gets used once. And I find that code questionable as well. If the only existence of the call goes away, this should also disapear

Ivanca replied:
Done

felipemontoya commented on seb_openedx/views.py:
```diff
@@ -24,10 +25,14 @@ def info_view(request):
     except CalledProcessError:
         git_data = ''
 
+    middlewares = settings.MIDDLEWARE_CLASSES if hasattr(settings, 'MIDDLEWARE_CLASSES') else settings.MIDDLEWARE
```
This whole function is cool for debugging, but its very own existence should be defaulted to false. The reasoning being that we don't want to let a potencial fraudster know which exact version of the code will answer to his/her requests.

Ivanca replied:
Done (now it works for staff only)

felipemontoya commented on setup.py:
```diff
@@ -42,5 +42,8 @@ def get_version(*file_paths):
         "lms.djangoapp": [
             "seb_openedx = seb_openedx.apps:SebOpenEdxConfig",  # retain
         ],
+        "cms.djangoapp": [
```
Last time we tried to have two endpoints with the same entry point, it failed. Is this working alright?

Ivanca replied:
Yeah It's working alright (but cms part is going to be deleted so it doesn't matter much)

felipemontoya reviewed (changes requested):
On a general note, I think this PR would be more elegant and simple if we:
- removed the whole CMS part
- Assumed the configurations are in where they need to be
- Moved most of the logic to classes we can mix and match in the configurations

Ivanca committed:
Removing CMS logic from the plugin (it handles lms only now)
- Removing enforcement of FEATURES['ENABLE_OTHER_COURSE_SETTINGS'] (should be set by administrator instead for plugin to work)
- always_allow_staff is now a decorator one may use to always allow staff users.

Ivanca committed:
Updating documentation to explain how to install and configure this thing

Ivanca commented:
@felipemontoya  Feedback changes done on last couple commits

felipemontoya commented on seb_openedx/middleware.py:
```diff
@@ -0,0 +1,20 @@
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        """ Start point of to d4etermine cms or lms """
+        if request.user.is_authenticated():
+            course_key_string = view_kwargs.get('course_key_string') or view_kwargs.get('course_id')
```
This seems to me like a weak implementation. What happens if the value is called course_key for any view? Could that even happen? Even if we check for the now common strings here, it still is just waiting for a dev to call it diferently and we break.

felipemontoya commented on seb_openedx/lms.py:
```diff
@@ -0,0 +1,18 @@
+class lms(object):
+
+    # pylint: disable=inconsistent-return-statements
+    @always_allow_staff
```
Can I chain decorators? Also, is a decorator the best way of doing this? What if I want to pass this over a long list of possible functions (decorators in this case)?

felipemontoya commented on seb_openedx/middleware.py:
```diff
@@ -0,0 +1,20 @@
+            course_key_string = view_kwargs.get('course_key_string') or view_kwargs.get('course_id')
+            course_key = CourseKey.from_string(course_key_string) if course_key_string else None
+            if course_key:
+                course_module = modulestore().get_course(course_key, depth=0)
```
What does depth=0 means? Only get the most global information?
Also, this dependency on `xmodule.modulestore.django` makes us depend on the edx-platform repo in order to even load this module in memory. Can we separate it in the style of eox?

felipemontoya commented on seb_openedx/views.py:
```diff
@@ -18,18 +19,24 @@ def info_view(request):
-        json.dumps(response_data),
-        content_type="application/json"
-    )
+    if request.user.is_authenticated() and request.user.is_staff:
```
I like :+1:

felipemontoya commented on seb_openedx/views.py:
```diff
@@ -18,18 +19,24 @@ def info_view(request):
+            git_data = ''
+
+        middlewares = settings.MIDDLEWARE_CLASSES if hasattr(settings, 'MIDDLEWARE_CLASSES') else settings.MIDDLEWARE
+        is_middleware_installed = 'seb_openedx.middleware.SecureExamBrowserMiddleware' in middlewares
```
is it possible to make this depend only of 'seb_openedx.middleware' ?

Ivanca replied:
Done at https://github.com/eduNEXT/seb-openedx/pull/1/commits/4feacb88c069bc1e5836df1e2c143bdb288fb4f3

Ivanca committed:
Making permissions as a list of classes, abstracting back-end (using wrapper)

Ivanca committed:
new is_installed class method

Ivanca commented:
@felipemontoya latest feedback addressed on https://github.com/eduNEXT/seb-openedx/pull/1/commits/f56b66e81011cd8e3606dd33911db6e4be91a742

Ivanca committed:
Fixing test: TestInfoView runs succesfully now

Ivanca commented:
Fixing test on https://github.com/eduNEXT/seb-openedx/pull/1/commits/300912228f77135c5233be7a0bee5f64c295235d It is adviced to approve this commit to continue in another

felipemontoya reviewed (approved):
Great work @Ivanca
